docs/new_protocol.asciidoc

```diff
@@ -0,0 +1,463 @@
+			// drop this tcp stream. Will retry parsing with the next
+			// segment in it
+			priv.Data[dir] = nil
+			logp.Debug("mysql", "Ignore MySQL message. Drop tcp stream.")
+			return priv
+		}
+
+		if complete {
+			mysql.messageComplete(tcptuple, dir, stream)
+		} else {
+			// wait for more data
+			break
+		}
+----------------------------------------------------------------------
+
+The `mysqlMessageParser()` is the function that tries to parse a single MySQL
+message. Its implementation is MySQL specific so not interesting to us for this
+guide. It returns two values: `ok` which is `false` if there was a parsing error
+from which we cannot recover and `complete` which indicates whether a complete
+and valid message was separated from the stream. These two values are used for
+deciding what to do next. In case of errors, we drop the stream. If there are no
+errors, but the message is not yet complete, we do nothing and wait for more
+data. Finally, if the message is complete, we go to the next level.
+
+This block of code is called in a loop so that it can separate multiple messages
+found in the same packet.
+
+==== The UDP ParseUdp function
+
+If the protocol you are working on is running on top of UDP, then all the
+complexities around extracting messages from packets that the TCP
+parser/decoders have to are not required at all.
+
+For an example, see the `ParseUdp()` function from the DNS module.
+
+==== Correlation
+
+Most protocols that Packetbeat supports today are request-response oriented.
+Packetbeat indexes into Elasticsearch a document for each request-response pair
+(called a transaction). This way we can have data from the request and the
+response in the same document and measure the response time.
+
+But this can be different for your protocol, for example for an asynchronous
+protocol like AMPQ, it makes more sense to index a document for every message,
+and then no correlation is necessary. On the other hand, for a session based
+protocol like SIP it might make sense to index a document for a SIP transaction
+or for a full SIP dialog, which can have more than two messages.
+
+The TCP stream or UDP ports are usually good indicators that two messages belong
+to the same transactions. Therefore most protocol implementations we have in
+Packatbeat use a map with `tcptuple` maps for correlating the requests with the
+responses.
+
+=== Testing
+
+==== Unit tests
+
+For unit tests, use only the Go standard library
+http://golang.org/pkg/testing/[testing] package. To make comparing complex
+structures less verbose, we use the assert package from the
+https://github.com/stretchr/testify[testify] library.
+
+For parser/decoder tests, we find it is a good practice to have an array with
+test cases containing the inputs and expected outputs. For an example, see for
+example the
+https://github.com/elastic/packetbeat/blob/b9173ae034581205ed4853c6fb040ea5357a5c28/protos/http/http_test.go#L1012[`Test_splitCookiesHeaders`]
+unit test.
```

Comment:
`gopacket` provides [test_creator.py](https://github.com/google/gopacket/blob/master/layers/test_creator.py) that creates `[]byte` from pcaps.